---
title: "Humann3 Report"
date: "`r format(Sys.Date(), '%Y-%m-%d')`"
header-includes:
- \usepackage{placeines}
- \usepackages{float}
- \floatplacement{figure}{H}

output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    df_print: paged
urlcolor: blue
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(include = TRUE)
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(cache = FALSE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(out.extra = '')
knitr::opts_chunk$set(as.result=TRUE)
knitr::opts_chunk$set(dev.args=list(png=list(type="cairo")))

# required r libraries
library(knitr)
library(ggplot2)
library(dplyr)
library(tidyr)
library(forcats)
library(DT)

```

```{r read-data}
path_coverage <- read.table("../humann3.dir/merged_pathcoverage.tsv", sep="\t")
path_abundance <- read.table("../humann3.dir/merged_pathabundance.tsv", sep="\t")
gene_abundance <- read.table("../humann3.dir/merged_genefamilies.tsv", sep="\t")
taxa_abundance <- list()
iter_file <- list.files("../metaphlan_output.dir/", pattern = "metaphlan_.*\\.tsv")
for(i in 1:length(iter_file)) {
  curr_level <- gsub("metaphlan_", "", iter_file[i])
  curr_level <- gsub(".tsv", "", curr_level)
  curr_file <- file.path("../metaphlan_output.dir", iter_file[i])
  taxa_abundance[[curr_level]] <- read.table(curr_file, sep="\t")
}

# update headers
header <- readLines("../humann3.dir/merged_pathcoverage.tsv", n=1)
header <- unlist(strsplit(header, '\t'))
header[1] <- gsub("^# ", "", tolower(header[1]))
header <- gsub("_pathcoverage.tsv$", "", header)
colnames(path_coverage) <- header

header <- readLines("../humann3.dir/merged_pathabundance.tsv", n=1)
header <- unlist(strsplit(header, '\t'))
header[1] <- gsub("^# ", "", tolower(header[1]))
header <- gsub("_pathabundance.tsv$", "", header)
colnames(path_abundance) <- header

header <- header <- readLines("../humann3.dir/merged_genefamilies.tsv", n=1)
header <- unlist(strsplit(header, '\t'))
header <- gsub("_genefamilies.tsv$", "", header)
header[1] <- gsub("^# ", "", header[1])
header[1] <- gsub(" ", "_", tolower(header[1]))
colnames(gene_abundance) <- header
```

## Pipeline parameters

```{r pipeline-parameters}
yml <- readLines("../pipeline.yml")
# only keep parameters
yml <- yml[grepl(":", yml)]
# remove unnecessary spaces at beginning
yml <- gsub("^ +", "", yml)

# parse humann3 options
humann3_option <- yml[grepl("db_nucleotide:|db_protein:|nthreads:|options:", yml)]
humann3_option <- sapply(humann3_option, function(x) strsplit(x, ": "))
humann3_entry <- c()
for(i in humann3_option) {
  curr <- data.frame(parameter = i[1], value = i[2])
  humann3_entry <- rbind(humann3_entry, curr)
}
humann3_entry$pipeline_task <- 'runHumann3'
humann3_entry <- humann3_entry %>% select(pipeline_task, everything())

# put into table
yml_table <- DT::datatable(humann3_entry, rownames=FALSE)
```

```{r yml-table}
yml_table
```

## Overview
```{r overview}
# number of samples
n_sample <- ncol(gene_abundance) - 1

# summarise of species observed
species <- gsub("^.*\\|", "", gene_abundance$gene_family)
unique_species <- unique(species[grepl("^g__", species)])
n_species <- length(unique_species)

# summarise gene families observed
gene_family <- gene_abundance$gene_family
unique_gene_family <- unique(gene_family[!grepl("\\|g__", gene_family)])
n_gene_family <- length(unique_gene_family)

# summarise pathways observed
pathway <- path_coverage$pathway
unique_pathway <- unique(pathway[!grepl("\\|g__", pathway)])
n_pathway <- length(unique_pathway)
```

Data set contains: 

* **Number of samples:** `r n_sample` samples
* **Number of species:** `r n_species` unique species
* **Number of gene families:** `r n_gene_family` unique gene families
* **Number of pathways:** `r n_pathway` unique pathways

## Unique taxa observed {.tabset}
### Kingdom
```{r taxa-kingdom}
DT::datatable(taxa_abundance$kingdom)
```

### Phylum
```{r taxa-phylum}
DT::datatable(taxa_abundance$phylum)
```

### Class
```{r taxa-class}
DT::datatable(taxa_abundance$class)
```

### Order
```{r taxa-order}
DT::datatable(taxa_abundance$order)
```

### Family
```{r taxa-family}
DT::datatable(taxa_abundance$family)
```

### Genus
```{r taxa-genus}
DT::datatable(taxa_abundance$genus)
```

### Species
```{r taxa-species}
DT::datatable(taxa_abundance$species)
```

## Gene Families: 
From Humann3 manual:

* This file details the abundance of each gene family in the community. Gene families are groups of evolutionarily-related protein-coding sequences that often perform similar functions.
* Gene family abundance at the community level is stratified to show the contributions from known and unknown species. Individual species' abundance contributions sum to the community total abundance.
* HUMAnN 3.0 uses the MetaPhlAn2 software along with the ChocoPhlAn database and translated search database for this computation.
* Gene family abundance is reported in RPK (reads per kilobase) units to normalize for gene length; RPK units reflect relative gene (or transcript) copy number in the community. RPK values can be further sum-normalized to adjust for differences in sequencing depth across samples.
* The "UNMAPPED" value is the total number of reads which remain unmapped after both alignment steps (nucleotide and translated search). Since other gene features in the table are quantified in RPK units, "UNMAPPED" can be interpreted as a single unknown gene of length 1 kilobase recruiting all reads that failed to map to known sequences.
* The UniRef50_unknown values represent the total abundance of reads which map to ChocoPhlAn nucleotide sequences which do not have a UniRef50 annotation.

### Community
Showing top 50 gene family abundance at the community level. Box plots depict interquartile range of samples. 

```{r gene-community}
avg_gene_abund <- gene_abundance %>%
  filter(gene_family %in% unique_gene_family) %>%
  gather('sample_name', 'abundance_RPK', -gene_family) %>%
  group_by(gene_family) %>%
  mutate(avg_abund_RPK = mean(abundance_RPK)) 

avg_gene <- avg_gene_abund %>% 
  distinct(gene_family, avg_abund_RPK)
genes_to_plot <- avg_gene %>%
  ungroup() %>%
  arrange(desc(avg_abund_RPK)) %>%
  slice(1:50) %>%
  pull(gene_family)

pdata <- avg_gene_abund %>% 
  filter(gene_family %in% genes_to_plot) %>%
  ungroup() %>%
  mutate(gene_family = fct_reorder(gene_family, avg_abund_RPK))

p_gene_community <- ggplot(pdata, aes(y = gene_family, x = abundance_RPK)) +
  stat_boxplot(fill=NA) +
  theme_bw(12)
```

```{r gene-community-plot, fig.height=18/2.54, fig.width=15/2.54}
p_gene_community
```

## Pathway Coverage
From Humann3 manual:

* Pathway coverage provides an alternative description of the presence (1) and absence (0) of pathways in a community, independent of their quantitative abundance.
* More specifically, HUMAnN 3.0 assigns a confidence score to each reaction detected in the community. Reactions with abundance greater than the median reaction abundance are considered to be more confidently detected than those below the median abundance.
* HUMAnN 3.0 then computes pathway coverage using the same algorithms described above in the context of pathway abundance, but substituting reaction confidence for reaction abundance.
* A pathway with coverage = 1 is considered to be confidently detected (independent of its abundance), as this implies that all of its member reactions were also confidently detected. A pathway with coverage = 0 is considered to less confidently detected (independent of its abundance), as this implies that some of its member reactions were not confidently detected.
* Like pathway abundance, pathway coverage is computed for the community as a whole, as well as for each detected species and the unclassified stratum.
Much as community-level pathway abundance is not the strict sum of species-level contributions, it is possible for a pathway to be confidently covered at the community level but never confidently detected from any single species.
* Pathway coverage is reported for any non-zero pathway abundance computed at the community-level or for an individual stratum (species or "unclassified").

### Community
Showing top 50 pathways at the community level. Box plots depict interquartile range of samples.

```{r pathcov-community}
avg_path_coverage <- path_coverage %>%
  filter(pathway %in% unique_pathway) %>%
  gather('sample_name', 'coverage', -pathway) %>%
  group_by(pathway) %>%
  mutate(avg_coverage = mean(coverage)) 

avg_path <- avg_path_coverage %>% 
  distinct(pathway, avg_coverage)
path_to_plot <- avg_path %>%
  ungroup() %>%
  arrange(desc(avg_coverage)) %>%
  slice(1:50) %>%
  pull(pathway)

pdata <- avg_path_coverage %>% 
  filter(pathway %in% path_to_plot) %>%
  ungroup() %>%
  mutate(pathway = fct_reorder(pathway, avg_coverage))

p_pathcov_community <- ggplot(pdata, aes(y = pathway, x = coverage)) +
  stat_boxplot(fill=NA) +
  theme_bw(12)
```

```{r pathcov-community-plot, fig.height=18/2.54, fig.width=25/2.54}
p_pathcov_community
```

## Pathway Abundance
From Humann3 manual:

* This file details the abundance of each pathway in the community as a function of the abundances of the pathway's component reactions, with each reaction's abundance computed as the sum over abundances of genes catalyzing the reaction.
* Pathway abundance is computed once at the community level and again for each species (plus the "unclassified" stratum) using community- and species-level gene abundances along with the structure of the pathway.
* The pathways are ordered by decreasing abundance with pathways for each species also sorted by decreasing abundance. Pathways with zero abundance are not included in the file.
* Pathway abundance is proportional to the number of complete "copies" of the pathway in the community. Thus, for a simple linear pathway RXN1â†’RXN2â†’RXN3â†’RXN4, if RXN1 is 10 times as abundant as RXNs 2-4, the pathway abundance will be driven by the abundances of RXNs 2-4.
* Unlike gene abundance, a pathway's community-level abundance is not necessarily the sum of its stratified abundance values. For example, continuing with the simple linear pathway example introduced above, if the abundances of RXNs 1-4 are [5, 5, 10, 10] in Species_A and [10, 10, 5, 5] in Species_B, HUMAnN 3.0 would report that Species_A and Species_B each contribute 5 complete copies of the pathway. However, at the community level, the reaction totals are [15, 15, 15, 15], and thus HUMAnN 3.0 would report 15 complete copies.
* In greater detail, the abundance for each pathway is a recursive computation of abundances of sub-pathways with paths resolved to abundances based on the relationships and abundances of the reactions contained in each. Each path, the smallest portion of a pathway or sub-pathway which can't be broken down into sub-pathways, has an abundance that is the max or harmonic mean of the reaction abundances depending on the relationships of these reactions. Optional reactions are only added to the overall abundance if their abundance is greater than the harmonic mean of the required reactions.
* Gap filling allows for a single required reaction to have a zero abundance. For all pathways, the required reaction with the lowest abundance is replaced with the abundance of the required reaction with the second lowest abundance.
* By default, HUMAnN 3.0 uses MetaCyc pathway definitions and MinPath to identify a parsimonious set of pathways which explain observed reactions in the community.
* To account for non-linearity in the conversion of gene copy number to pathway copy number, we define a "compression constant" (k) equal to the total pathway abundance divided by the total abundance of genes that contributed to pathways. The "UNMAPPED" value reported in the pathway abundance table is equal to the total number of unmapped reads scaled by k (making it more comparable with pathway abundance values). Similarly, we define an "UNINTEGRATED" abundance for 1) the community, 2) each identified species, and 3) the "unclassified" stratum equal to the total abundance of genes in that level that did not contribute to pathways (scaled by k).
* "UNINTEGRATED" does not appear for stratifications with no detected pathways.

### Community
Showing top 50 pathways at the community level. Box plots depict interquartile range of samples.

```{r pathabund-community}
avg_path_abund <- path_abundance %>%
  filter(pathway %in% unique_pathway) %>%
  gather('sample_name', 'abundance', -pathway) %>%
  group_by(pathway) %>%
  mutate(avg_abundance = mean(abundance)) 

avg_path <- avg_path_abund %>% 
  distinct(pathway, avg_abundance)
path_to_plot <- avg_path %>%
  ungroup() %>%
  arrange(desc(avg_abundance)) %>%
  slice(1:50) %>%
  pull(pathway)

pdata <- avg_path_abund %>% 
  filter(pathway %in% path_to_plot) %>%
  ungroup() %>%
  mutate(pathway = fct_reorder(pathway, avg_abundance))

p_pathabund_community <- ggplot(pdata, aes(y = pathway, x = abundance)) +
  stat_boxplot(fill=NA) +
  theme_bw(12)
```

```{r pathabund-community-plot, fig.height=18/2.54, fig.width=25/2.54}
p_pathabund_community
```